seerep_srv/seerep_server/src/fb_image_service.cpp

```diff
@@ -110,37 +110,29 @@ grpc::Status FbImageService::TransferImage(
   flatbuffers::grpc::Message<seerep::fb::Image> imageMsg;
   std::vector<std::pair<std::string, boost::uuids::uuid>> projectImgUuids;
 
-  // add incoming image data to hdf5
   while (reader->Read(&imageMsg))
   {
-    BOOST_LOG_SEV(m_logger, boost::log::trivial::severity_level::info)
-        << "received image... ";
-    auto image = imageMsg.GetRoot();
-    std::string uuidProject = image->header()->uuid_project()->str();
+    const seerep::fb::Image* image = imageMsg.GetRoot();
+
+    if ((image->uri() == nullptr || image->uri()->str().empty()) &&
+        image->data() == nullptr)
+    {
+      answer = "Image message must contain an uri or payload";
+      return grpc::Status(grpc::StatusCode::INVALID_ARGUMENT, answer);
+    }
+
+    const std::string projectUuid = image->header()->uuid_project()->str();
 
-    if (!uuidProject.empty())
+    if (!projectUuid.empty())
     {
       try
       {
         projectImgUuids.push_back(
-            { uuidProject, imageFb->addDataToHdf5(*image) });
-      }
-      catch (std::runtime_error const& e)
-      {
-        // mainly catching "invalid uuid string" when transforming uuid_project
-        // from string to uuid also catching core doesn't have project with uuid
-        // error
-        BOOST_LOG_SEV(m_logger, boost::log::trivial::severity_level::error)
-            << e.what();
-        seerep_server_util::createResponseFb(
-            std::string(e.what()), seerep::fb::TRANSMISSION_STATE_FAILURE,
-            response);
-        return grpc::Status(grpc::StatusCode::INVALID_ARGUMENT, e.what());
+            { projectUuid, imageFb->addDataToHdf5(*image) });
+        imageFb->buildIndices(projectImgUuids);
```

Comment:
We want to do the buildIndices outside of the `while (reader->Read(&imageMsg))` so that the streaming is as fast as possible. Thus, the buildIndices was in a separate try/catch

Reply:
I don't think this will make the overall stream faster. The while loop handling the incoming images will be faster. However, when building the indices, data gets read from the disk again:
https://github.com/agri-gaia/seerep/blob/b3f61d8293811a182a90257855723e81dbcc84f7/seerep_srv/seerep_core_fb/src/core_fb_image.cpp#L86-L87
which will take a considerable amount of time that gets added before the response is returned to the client (using `gRPC::Status::Ok` is this case).

Reply:
No, not the overall stream, but we can handle a higher frequency of images as we do not have to build the inidces for each image before receiving another

Reply:
Ah, ok. Did anyone measure the difference?

Reply:
> Ah, ok. Did anyone measure the difference?

I added my findings while I implemented that feature [here](https://github.com/agri-gaia/seerep/pull/385#issue-2369805984).